### PR TITLE
skiplist: not use CAS for single thread writing

### PR DESCRIPTION
**Pull request by zhangjinpeng1987**
Signed-off-by: zhangjinpeng1987 <zhangjinpeng@pingcap.com>

### CODE REVIEW

**Review by skyzh — COMMENTED** [review]
In what cases will we concurrently write to memtable?

https://github.com/tikv/agatedb/blob/366404d0a4960eeddd8e5dfd1af318716e621cfd/src/db.rs#L248-L254

Currently, all writes go into only one thread.

**Comment by zhangjinpeng1987** [comment]
> In what cases will we concurrently write to memtable?
> 
> https://github.com/tikv/agatedb/blob/366404d0a4960eeddd8e5dfd1af318716e621cfd/src/db.rs#L248-L254
> 
> Currently, all writes go into only one thread.

@skyzh The context is that, skiplist's previous implementation works under both single-thread write and multi-threads write. But there is a more efficient way to implement for single thread write scenario which just use atomic.store, don't use CAS.

> Currently, all writes go into only one thread.
This is true, but skiplist is a fundamental component, maybe in the future we need the concurrent write ability.

File: skiplist/src/list.rs [hunk 1]
```diff
@@ -223,7 +225,7 @@ impl<C: KeyComparator> Skiplist<C> {
         }
```
Comment (GanZiheng) [inline skiplist/src/list.rs:225]: Should we also change the logic of increasing height if we add `allow_concurrent_write` option?

Comment (zhangjinpeng1987) [reply skiplist/src/list.rs:232]: done

**Comment by zhangjinpeng1987** [comment]
@skyzh Any other concerns or comments?

**Review by skyzh — APPROVED** [review]

**Review by GanZiheng — APPROVED** [review]
LGTM

File: skiplist/src/list.rs [hunk 2]
```diff
@@ -210,20 +212,68 @@ impl<C: KeyComparator> Skiplist<C> {
...
+                    height,
+                    Ordering::SeqCst,
+                    Ordering::SeqCst,
+                ) {
+                    Ok(_) => break,
+                    Err(h) => list_height = h,
+                }
+            }
+        } else {
+            // There is no need to use CAS for single thread writing.
```
Comment (Connor1996) [inline skiplist/src/list.rs:228]: we plan to support the multiple thread writing, it's necessary now? We may need to delete in future.

Comment (zhangjinpeng1987) [reply skiplist/src/list.rs:228]: If treat AgateDB as a fundamental kv engine on cloud, maybe it need both single thread write and multi-threads write.

**Pull request merged**